embassy-rp/src/trng.rs

```diff
@@ -0,0 +1,428 @@
+            return; // Nothing to fill
+        }
+
+        self.start_rng();
+        self.enable_irq();
+
+        let mut bytes_transferred = 0usize;
+        let mut buffer = [0u8; TRNG_BLOCK_SIZE_BYTES];
+
+        let regs = T::regs();
+
+        let trng_busy_register = regs.trng_busy();
+        let trng_valid_register = regs.trng_valid();
+
+        let waker = T::waker();
+
+        let destination_length = destination.len();
+
+        let mut counter = 0;
+        poll_fn(|context| {
+            critical_section::with(|_cs| {
+                counter += 1;
+                waker.register(context.waker());
+                if bytes_transferred == destination_length {
+                    self.stop_rng();
+                    self.disable_irq();
+                    Poll::Ready(())
+                } else {
+                    if trng_busy_register.read().trng_busy() {
+                        Poll::Pending
+                    } else {
+                        if trng_valid_register.read().ehr_valid().not() {
+                            panic!("RNG not busy, but ehr is not valid!")
+                        }
+                        self.read_ehr_registers_into_array(&mut buffer);
+                        let remaining = destination_length - bytes_transferred;
+                        self.blocking_read_ehr_registers_into_array(&mut buffer);
+                        match remaining {
+                            size if size > TRNG_BLOCK_SIZE_BYTES => {
+                                destination[bytes_transferred..bytes_transferred + TRNG_BLOCK_SIZE_BYTES]
+                                    .copy_from_slice(&buffer);
+                                bytes_transferred += TRNG_BLOCK_SIZE_BYTES
+                            }
+                            _ => {
+                                destination[bytes_transferred..bytes_transferred + remaining]
+                                    .copy_from_slice(&buffer[0..remaining]);
+                                bytes_transferred += remaining
+                            }
+                        }
+                        if bytes_transferred == destination_length {
+                            self.stop_rng();
+                            self.disable_irq();
+                            Poll::Ready(())
+                        } else {
+                            Poll::Pending
+                        }
+                    }
+                }
+            })
+        })
+        .await
+    }
+
+    /// Fill the buffer with random bytes, blocking version.
+    pub fn blocking_fill_bytes(&mut self, destination: &mut [u8]) {
+        if destination.is_empty() {
+            return; // Nothing to fill
+        }
+        self.start_rng();
+
+        let destination_length = destination.len();
+        let mut bytes_transferred = 0usize;
+
+        let mut buffer = [0u8; TRNG_BLOCK_SIZE_BYTES];
+
+        while bytes_transferred < destination_length {
```

Comment:
this code is a bit hard to follow. You're doing chunking manually, you have a  `match remaining` that could be written as an `if`.

I'd suggest chunking with `for chunk in destination.chunks(TRNG_BLOCK_SIZE_BYTES)`, then you just have to deal with the last chunk being shorter than TRNG_BLOCK_SIZE_BYTES.

Reply:
with `chunk.copy_from_slice(&buffer[..chunk.len()])`